File: services/customers-service/src/main/java/br/com/deltaglobalbank/customers/features/customers/createCustomer/CreateCustomerUseCase.java
package br.com.deltaglobalbank.customers.features.customers.createCustomer;

import java.util.List;
import java.util.UUID;

import br.com.deltaglobalbank.customers.domain.bankAccount.BankAccount;
import br.com.deltaglobalbank.customers.domain.bankAccount.BankAccountPurpose;
import br.com.deltaglobalbank.customers.domain.bankAccount.BankAccountType;
import br.com.deltaglobalbank.customers.domain.bankAccount.valueobjects.Agency;
import br.com.deltaglobalbank.customers.domain.bankAccount.valueobjects.BankCode;
import br.com.deltaglobalbank.customers.domain.customer.CpfAlreadyExists;
import br.com.deltaglobalbank.customers.domain.customer.Customer;
import br.com.deltaglobalbank.customers.domain.customer.CustomerRepository;
import br.com.deltaglobalbank.customers.domain.customer.CustomerSnapshot;
import br.com.deltaglobalbank.customers.domain.customer.valueobjects.Address;
import br.com.deltaglobalbank.customers.domain.customer.valueobjects.BirthDate;
import br.com.deltaglobalbank.customers.domain.customer.valueobjects.Cpf;
import br.com.deltaglobalbank.customers.domain.customer.valueobjects.Email;
import br.com.deltaglobalbank.customers.domain.customer.valueobjects.FullName;
import br.com.deltaglobalbank.customers.domain.customer.valueobjects.Gender;
import br.com.deltaglobalbank.customers.domain.customer.valueobjects.MaritalStatus;
import br.com.deltaglobalbank.customers.domain.customer.valueobjects.MotherName;
import br.com.deltaglobalbank.customers.domain.customer.valueobjects.Phone;
import br.com.deltaglobalbank.customers.domain.document.DocumentType;
import br.com.deltaglobalbank.customers.domain.document.PersonalDocument;
import br.com.deltaglobalbank.customers.domain.shared.valueobjects.Uf;
import br.com.deltaglobalbank.customers.infrastructure.persistence.entities.DocumentIssuerEntity;
import br.com.deltaglobalbank.customers.infrastructure.persistence.repositories.JpaDocumentIssuerRepository;
import com.github.f4b6a3.uuid.UuidCreator;
import jakarta.transaction.Transactional;
import org.springframework.stereotype.Service;

@Service
public class CreateCustomerUseCase {

    private final CustomerRepository customerRepository;
    private final JpaDocumentIssuerRepository jpaDocumentIssuerRepository;

    public CreateCustomerUseCase(
        CustomerRepository customerRepository,
        JpaDocumentIssuerRepository jpaDocumentIssuerRepository
    ) {
        this.customerRepository = customerRepository;
        this.jpaDocumentIssuerRepository = jpaDocumentIssuerRepository;
    }

    @Transactional
    public CreateCustomerResponse execute(CreateCustomerCommand command) {
        CreateCustomerRequest request = command.request();
        Cpf cpf = new Cpf(request.cpf());
        if (customerRepository.existsByCpfAndTenantId(cpf, command.tenantId())) {
            throw new CpfAlreadyExists();
        }

        Gender gender = Gender.fromDatabaseValue(request.gender());
        MaritalStatus maritalStatus = MaritalStatus.fromDatabaseValue(request.maritalStatus());
        Email email = request.email() != null ? new Email(request.email()) : null;
        Phone phone = new Phone(request.phone().phoneNumber());
        Address address = new Address(
            request.address().cep(),
            request.address().street(),
            request.address().city(),
            new Uf(request.address().state()),
            request.address().country(),
            request.address().number(),
            request.address().complement(),
            request.address().neighborhood()
        );

        Customer customer = Customer.create(
            UuidCreator.getTimeOrderedEpoch(),
            command.tenantId(),
            cpf,
            new FullName(request.fullName()),
            new BirthDate(request.birthDate()),
            gender,
            request.nationality(),
            new MotherName(request.motherName()),
            maritalStatus,
            email,
            phone,
            address,
            command.createdBy(),
            List.of(),
            List.of()
        );

        for (BankAccountRequest acc : request.bankAccounts()) {
            BankAccount account = BankAccount.create(
                UuidCreator.getTimeOrderedEpoch(),
                customer.id(),
                new BankCode(acc.bankCode()),
                new Agency(acc.agency()),
                acc.accountNumber(),
                acc.accountDigit(),
                BankAccountType.fromDatabaseValue(acc.accountType()),
                BankAccountPurpose.fromDatabaseValue(acc.purpose()),
                acc.isPrimary()
            );
            customer.addBankAccount(account, command.createdBy());
        }

        List<PersonalDocument> documents = request.documents().stream()
            .map(doc -> {
                DocumentIssuerEntity issuer = jpaDocumentIssuerRepository.findByName(doc.issuer());
                if (issuer == null) {
                    throw new IllegalArgumentException("invalid_issuer");
                }
                return PersonalDocument.create(
                    UuidCreator.getTimeOrderedEpoch(),
                    customer.id(),
                    DocumentType.fromDatabaseValue(doc.type()),
                    doc.number(),
                    issuer.getId(),
                    new Uf(doc.issuerState()),
                    null,
                    doc.issuedAt()
                );
            })
            .toList();
        customer.replaceDocuments(documents, command.createdBy());

        Customer saved = customerRepository.save(customer);

        CustomerSnapshot s = saved.snapshot();
        return new CreateCustomerResponse(
            s.id(),
            s.tenantId(),
            s.cpf().value(),
            s.fullName().value(),
            s.birthDate().value(),
            s.status().toDatabaseValue(),
            s.createdAt()
        );
    }
}

File: services/customers-service/src/main/kotlin/br/com/deltaglobalbank/customers/features/customers/createCustomer/CreateCustomerUseCase.kt
package br.com.deltaglobalbank.customers.features.customers.createCustomer

import br.com.deltaglobalbank.customers.domain.bankAccount.BankAccount
import br.com.deltaglobalbank.customers.domain.bankAccount.BankAccountPurpose
import br.com.deltaglobalbank.customers.domain.bankAccount.BankAccountType
import br.com.deltaglobalbank.customers.domain.bankAccount.valueobjects.Agency
import br.com.deltaglobalbank.customers.domain.bankAccount.valueobjects.BankCode
import br.com.deltaglobalbank.customers.domain.customer.CpfAlreadyExists
import br.com.deltaglobalbank.customers.domain.customer.Customer
import br.com.deltaglobalbank.customers.domain.customer.CustomerRepository
import br.com.deltaglobalbank.customers.domain.customer.valueobjects.Address
import br.com.deltaglobalbank.customers.domain.customer.valueobjects.BirthDate
import br.com.deltaglobalbank.customers.domain.customer.valueobjects.Cpf
import br.com.deltaglobalbank.customers.domain.customer.valueobjects.Email
import br.com.deltaglobalbank.customers.domain.customer.valueobjects.FullName
import br.com.deltaglobalbank.customers.domain.customer.valueobjects.Gender
import br.com.deltaglobalbank.customers.domain.customer.valueobjects.MaritalStatus
import br.com.deltaglobalbank.customers.domain.customer.valueobjects.MotherName
import br.com.deltaglobalbank.customers.domain.customer.valueobjects.Phone
import br.com.deltaglobalbank.customers.domain.document.DocumentType
import br.com.deltaglobalbank.customers.domain.document.PersonalDocument
import br.com.deltaglobalbank.customers.domain.shared.valueobjects.Uf
import br.com.deltaglobalbank.customers.infrastructure.persistence.repositories.JpaDocumentIssuerRepository
import com.github.f4b6a3.uuid.UuidCreator
import jakarta.transaction.Transactional
import org.springframework.stereotype.Service
import java.util.UUID

data class CreateCustomerCommand(
    val tenantId: UUID,
    val createdBy: UUID,
    val request: CreateCustomerRequest,
)

@Service
class CreateCustomerUseCase(
    private val customerRepository: CustomerRepository,
    private val jpaDocumentIssuerRepository: JpaDocumentIssuerRepository,
) {
    @Transactional
    fun execute(command: CreateCustomerCommand): CreateCustomerResponse {
        val request = command.request
        val cpf = Cpf(request.cpf)
        if (customerRepository.existsByCpfAndTenantId(cpf, command.tenantId))
            throw CpfAlreadyExists()

        val gender = Gender.fromDatabaseValue(request.gender)
        val maritalStatus = MaritalStatus.fromDatabaseValue(request.maritalStatus)
        val email = request.email?.let { Email(it) }
        val phone = Phone(request.phone.phoneNumber)
        val address = Address(
            cep = request.address.cep,
            street = request.address.street,
            city = request.address.city,
            state = Uf(request.address.state),
            country = request.address.country,
            number = request.address.number,
            complement = request.address.complement,
            neighborhood = request.address.neighborhood,
        )

        val customer = Customer.create(
            id = UuidCreator.getTimeOrderedEpoch(),
            tenantId = command.tenantId,
            cpf = cpf,
            fullName = FullName(request.fullName),
            birthDate = BirthDate(request.birthDate),
            gender = gender,
            motherName = MotherName(request.motherName),
            maritalStatus = maritalStatus,
            email = email,
            phoneNumber = phone,
            address = address,
            createdBy = command.createdBy,
            nationality = request.nationality,
            bankAccounts = emptyList(),
            documents = emptyList(),

        )
        request.bankAccounts.forEach { acc ->
            val account = BankAccount.create(
                id = UuidCreator.getTimeOrderedEpoch(),
                customerId = customer.id,
                bankCode = BankCode(acc.bankCode),
                agency = Agency(acc.agency),
                accountNumber = acc.accountNumber,
                accountDigit = acc.accountDigit,
                accountType = BankAccountType.fromDatabaseValue(acc.accountType),
                purpose = BankAccountPurpose.fromDatabaseValue(acc.purpose),
                isPrimary = acc.isPrimary,
            )
            customer.addBankAccount(account, command.createdBy)   // lança DuplicatePrimaryAccountForPurpose → 400
        }

        val documents = request.documents.map { doc ->
            val issuerId = jpaDocumentIssuerRepository.findByName(doc.issuer)?.id
                ?: throw IllegalArgumentException("invalid_issuer")   // 400
            PersonalDocument.create(
                id = UuidCreator.getTimeOrderedEpoch(),
                customerId = customer.id,
                documentType = DocumentType.fromDatabaseValue(doc.type),
                documentNumber = doc.number,
                issuerId = issuerId,
                issuerState = Uf(doc.issuerState),
                expiresAt = null,
                issuedAt = doc.issuedAt,
            )
        }
        customer.replaceDocuments(documents, command.createdBy)

        val saved = customerRepository.save(customer)

        val s = saved.snapshot()
        return CreateCustomerResponse(
            id = s.id, tenantId = s.tenantId,
            cpf = s.cpf.value, fullName = s.fullName.value,
            birthDate = s.birthDate.value,
            status = s.status.toDatabaseValue(),
            createdAt = s.createdAt,
        )
    }
}

